script/run-in-docker.sh

```diff
@@ -7,5 +7,5 @@ image=fluree/${PWD##*/}
 echo "Running in ${image} container..."
 
 export DOCKER_BUILDKIT=1
-docker build --quiet --load --tag "${image}" .
+# docker build --cache-from type=gha --cache-to type=gha,mode=max --quiet --load --tag "${image}" .
```

Comment:
Hmm... I'm not sure about this change. Part of the purpose of this script is to lean on Docker caching to make builds fast(-ish) when little has changed but update them when things have changed, without having to worry about when the last time you built the ephemeral run-in-docker image. How much does "building" mostly / entirely from cache slow it down?

It's also intended to work locally too to provide very similar environments in both local dev and CI, so the cache type setting should maybe be behind a check for a "are we running in GitHub Actions?" env var or similar.

Reply:
> How much does "building" mostly / entirely from cache slow it down?

I'm not sure I understand the question? But will try to answer

I'm unsure about the degree to which the builds had been using caching, as I believe GHA would throw out the layers between jobs. In a local context, yeah we shouldn't have to worry about the cache flags at all. I think

In this line, I had attempted to modify `run-in-docker.sh` to start caching the docker layers, but it didn't appear to be working. It might actually work, I was changing many things at once between commits.

However I'm confident that using the `docker/build-push-action@v3` action in `ci.yml` is successfully caching the docker layers. I get your point though about this script no longer working for local builds.

Reply:
Seems like you understood. :) Basically asking if most or all of the Docker layers are already cached, how long does the build command take? The intent / hope was that it would be quick enough to just run every time.

Seems like you got caching working in GHA, and if we conditionalize it so it works locally too, hopefully we can re-enable the build command without a big perf hit. Happy to entertain alternative approaches that achieve similar goals, though.

I'm also happy to play around w/ it if you're ready to move on to something else (the YAML and GHA slow feedback loop fatigue is real!). :)

Reply:
I think I've hit my limit with 😅 But also I do need to focus on devrel stuff.

Re: build time with layers cached, I experimented with this by updating `ci.yml` to only include two identical jobs. Both jobs used the `docker/build-push-action@v3` action, and both had a `run` step of `run-in-docker.sh ls`, which I reasoned would reveal how much time is spent by the docker build in the presence of caching and in the absence of any additional work. Here's the result: https://github.com/fluree/db/actions/runs/3741708247

The result is a little misleading in that I had re-run the job, and it's only showing the results of the second run. In this second run, the first job was able to reuse the cache from the previous run, which is why it took only 1 minute. The results for the first run, which I'm not sure how to share, showed that the first job in that workflow took about 3 minutes.

You can see a break down of what gets cached and what gets rebuilt every time at https://github.com/fluree/db/actions/runs/3741708247/jobs/6353253651

Hope that helps!